### PR TITLE
[Inductor] FlexAttention supports block sparse mask

### PR DESCRIPTION
**Pull request by yanboliang**
Benchmark script (causal mask): https://gist.github.com/yanboliang/c2010a1fd081d4e8ca94fadec9eef286
Initial perf number:
* fwd speedup: 0.44 -> 0.72
* bwd speedup: 0.38 -> 0.71

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_dynamo/variables/higher_order_ops.py [hunk 1]
```diff
@@ -1559,7 +1577,7 @@ def call_function(
             proxy=tx.output.create_proxy(
                 "call_function",
                 self.value,
-                args=inp_args + p_args,
+                args=inp_args + p_args[:1] + sparse_mask_args + p_args[1:],
```
Comment (yanboliang) [inline torch/_dynamo/variables/higher_order_ops.py:1580]: TODO: sort these args in an organized way.

File: torch/_inductor/kernel/flex_attention.py [hunk 2]
```diff
@@ -199,6 +202,19 @@ def build_subgraph_buffer(
...
     q_offset = off_hz * stride_qh
     kv_offset = off_hz * stride_kh
+
+    BLOCKSPARSE_Q_MULTIPLE: tl.constexpr = (BLOCKSPARSE_Q // BLOCK_M)
+    BLOCKSPARSE_KV_MULTIPLE: tl.constexpr = (BLOCKSPARSE_KV // BLOCK_N)
+
+    # BLOCKSPARSE_Q_LEN: tl.constexpr = Q_LEN // BLOCKSPARSE_Q
+    BLOCKSPARSE_KV_LEN: tl.constexpr = KV_LEN // BLOCKSPARSE_KV
+
+    indices_offset = (start_m // BLOCKSPARSE_Q_MULTIPLE) * BLOCKSPARSE_KV_LEN
```
Comment (yanboliang) [inline torch/_inductor/kernel/flex_attention.py:212]: TODO: we should use stride rather than hardcoded ```BLOCKSPARSE_KV_LEN``` and support 4d mask, currently making mask tensor 2D for more efficient debugging.

File: torch/_inductor/kernel/flex_attention.py [hunk 3]
```diff
@@ -263,29 +282,40 @@ def build_subgraph_buffer(
...
+        acc *= acc_scale[:, None]
+        acc = tl.dot(p.to(MATMUL_PRECISION), v.to(MATMUL_PRECISION), acc)
+
+        # -- update m_i and l_i --
         l_i = l_i * alpha + tl.sum(p, 1)
-        # # -- scale and update acc --
-        acc = acc * alpha[:, None]
-        v = tl.load(V_block_ptr)
-        acc = tl.dot(p.to(MATMUL_PRECISION), v, acc)
+        m_i = m_i_new
```
Comment (yanboliang) [inline torch/_inductor/kernel/flex_attention.py:302]: L285-L302 is not from my change, it's from #129013 which has been reverted after my rebasing.

File: torch/_inductor/kernel/flex_attention.py [hunk 4]
```diff
@@ -425,10 +465,28 @@ def flex_attention(*args, **kwargs):
     # Note, we don't need to pass in the captured buffers explicitly
     # because they're implicitly added by the score_mod function
     # We do need to explicitly pass it in for autotuning though.
+
+    q_len = query.get_size()[-2]
+    kv_len = key.get_size()[-2]
+    sparse_mask_q_blocks = sparse_mask_kv_indices.get_size()[0]
+    sparse_mask_kv_blocks = sparse_mask_q_indices.get_size()[0]
+    BLOCKSPARSE_Q = q_len // sparse_mask_q_blocks
+    BLOCKSPARSE_KV = kv_len // sparse_mask_kv_blocks
```
Comment (yanboliang) [inline torch/_inductor/kernel/flex_attention.py:474]: TODO: This doesn't work well with dynamic shape, since either q_len or kv_len can be symbolic. ```BLOCKSPARSE_Q``` and ```BLOCKSPARSE_KV``` should be a configuration. This is just for backward test compatibility.

File: torch/nn/attention/_flex_attention.py [hunk 5]
```diff
@@ -39,11 +39,69 @@ def _identity(
...
+        q_indices,
+    ):
+        self.kv_num_blocks = kv_num_blocks
+        self.kv_indices = kv_indices
+        self.q_num_blocks = q_num_blocks
+        self.q_indices = q_indices
+
+
+def _create_block_sparse_mask(
+    mask: torch.Tensor,
```
Comment (yanboliang) [inline torch/nn/attention/_flex_attention.py:62]: TODO: making mask tensor 4D, currently I'm using 2D mask for debugging more simple.

File: torch/_inductor/kernel/flex_attention.py [hunk 6]
```diff
@@ -263,29 +282,40 @@ def build_subgraph_buffer(
...
 
-        # -- update m_i
-        m_i = m_ij
         # update pointers
-        K_block_ptr = tl.advance(K_block_ptr, (0, BLOCK_N))
-        V_block_ptr = tl.advance(V_block_ptr, (BLOCK_N, 0))
+        # if start_n < hi - 1:
+        indices_idx = start_n // BLOCKSPARSE_KV_MULTIPLE
+
+        cur_block = tl.load(kv_indices + indices_idx)
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_attention.py:308]: I know this was the original code but I think this is quite janky haha.

cc: @chenyang78 if you have any suggestions on how to write this code nicer.

File: torch/_inductor/kernel/flex_attention.py [hunk 7]
```diff
@@ -294,8 +324,8 @@ def build_subgraph_buffer(
     idx_m = offs_m[:, None]
     idx_d = tl.arange(0, BLOCK_DMODEL)[None, :]
 
-    mask = idx_m < Q_LEN
     # TODO generalize and add proper mask support
+    mask = (idx_m != -1) & (idx_d != -1)
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_attention.py:328]: change it back :P

File: torch/nn/attention/_flex_attention.py [hunk 8]
```diff
@@ -93,11 +151,22 @@ def score_mod(
...
+        block_sparse_mask = _create_empty_block_sparse_mask(query.device)
     if torch.compiler.is_dynamo_compiling():
         # mark head_dim always to be static
         for x in [query, key, value]:
             torch._dynamo.mark_static(x, -1)
-        out, _ = flex_attention_hop(query, key, value, score_mod)
+        out, _ = flex_attention_hop(
+            query,
+            key,
+            value,
```
Comment (drisspg) [inline torch/nn/attention/_flex_attention.py:163]: do we need to flatten the mask at this level?

File: torch/nn/attention/_flex_attention.py [hunk 9]
```diff
@@ -39,11 +39,69 @@ def _identity(
...
+    assert mask.dtype == torch.bool
+    q_len, kv_len = mask.shape
+    assert q_len % q_block_size == 0
+    assert kv_len % kv_block_size == 0
+    mask = mask.view(
+        q_len // q_block_size, q_block_size, kv_len // kv_block_size, kv_block_size
+    )
+
+    mask = mask.permute(0, 2, 1, 3)
+    block_mask = (mask.sum(dim=[-2, -1]) > 0).to("cpu")
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:75]: Change the dtype to `torch.int` instead of `torch.bool`

File: torch/nn/attention/_flex_attention.py [hunk 10]
```diff
@@ -39,11 +39,69 @@ def _identity(
     return score
 
 
+class _BlockSparseMask:
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:42]: Add BLOCKSPARSE_KV and BLOCKSPARSE_Q

**Review by Chillee — COMMENTED** [review]
Test a case where the blocksparsemask is compiled as part of the same function as flexattention.

```
@torch.compile
def f(q, k, v, mask):
    block_mask = compute_mask_info(mask)
    return flex_attention(q, k, v, block_mask)

@torch.compile
def f(q, k, v, mask):
    block_mask = compute_mask_info(mask)
    q = flex_attention(q, k, v, block_mask)
    return flex_attention(q, k2, v2, block_mask)

def f(q, k, v, block_mask):
    return flex_attention(q, k, v, block_mask)
```

File: torch/_dynamo/variables/higher_order_ops.py (hunk 1 above)
Comment (yanboliang) [reply torch/_dynamo/variables/higher_order_ops.py]: Done

File: torch/_inductor/kernel/flex_attention.py (hunk 2 above)
Comment (yanboliang) [reply torch/_inductor/kernel/flex_attention.py]: Done

File: torch/nn/attention/_flex_attention.py (hunk 5 above)
Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py]: Done

File: torch/_inductor/kernel/flex_attention.py (hunk 7 above)
Comment (yanboliang) [reply torch/_inductor/kernel/flex_attention.py]: Done

File: torch/nn/attention/_flex_attention.py (hunk 8 above)
Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py:213]: I think HOO doesn't support arbitrary argument types, so we have to flatten?

File: torch/nn/attention/_flex_attention.py (hunk 10 above)
Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py:45]: Done

File: test/inductor/test_flex_attention.py [hunk 11]
```diff
@@ -1008,13 +1165,21 @@ def forward(self, L_args_0_: "f64[2, 2, 8, 4]", L_args_1_: "f64[2, 2, 8, 4]", L_
         l_args_1_ = L_args_1_
         l_args_2_ = L_args_2_
 
+        ones: "i32[1, 1, 1]" = torch.ones([1, 1, 1], dtype = torch.int32, device = device(type='cuda', index=0))
```
Comment (Chillee) [inline test/inductor/test_flex_attention.py:1168]: This is starting to get quite ugly. I think we should figure out how to fix these in HOPs. cc: @zou3519 @ydwu4

File: torch/_dynamo/variables/higher_order_ops.py [hunk 12]
```diff
@@ -1581,7 +1602,7 @@ def call_function(
             proxy=tx.output.create_proxy(
                 "call_function",
                 self.value,
-                args=inp_args + p_args,
+                args=inp_args[:3] + p_args[:1] + inp_args[3:] + p_args[1:],
```
Comment (Chillee) [inline torch/_dynamo/variables/higher_order_ops.py:1605]: Why does this need to be done?

File: torch/_inductor/kernel/flex_attention.py [hunk 13]
```diff
@@ -156,7 +184,7 @@ def build_subgraph_buffer(
     name="flex_attention",
     grid=flex_attention_grid,
     source=r"""
-{{def_kernel("Q", "K", "V", "LSE")}}
+{{def_kernel("Q", "K", "V", "LSE", "SM_KV_NUM_BLKS", "SM_KV_IDX")}}
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_attention.py:187]: SM? Assuming it's not streaming-multiprocessor, it's kind of a confusing name to me.

File: torch/_inductor/kernel/flex_attention.py [hunk 14]
```diff
@@ -235,18 +289,21 @@ def build_subgraph_buffer(
...
+    hi = sm_kv_num_blocks * BLOCKSPARSE_KV_MULTIPLE
+
+    for start_n in range(0, hi):
         # -- load k, v --
         k = tl.load(K_block_ptr)
+        v = tl.load(V_block_ptr)
         # -- compute qk ---
-        qk = tl.dot(q, k)
+        qk = tl.zeros([BLOCK_M, BLOCK_N], dtype=tl.float32)
+        qk = tl.dot(q, k.to(MATMUL_PRECISION), acc=qk)
```
Comment (Chillee) [inline torch/_inductor/kernel/flex_attention.py:303]: Don't think we need to revert this.

File: torch/nn/attention/_flex_attention.py [hunk 15]
```diff
@@ -39,11 +39,119 @@ def _identity(
...
+        kv_num_blocks=kv_num_blocks.to(torch.int32).to(mask.device).contiguous(),
+        kv_indices=kv_indices.to(torch.int32).to(mask.device).contiguous(),
+        q_num_blocks=q_num_blocks.to(torch.int32).to(mask.device).contiguous(),
+        q_indices=q_indices.to(torch.int32).to(mask.device).contiguous(),
+        BLOCKSPARSE_KV=BLOCKSPARSE_KV,
+        BLOCKSPARSE_Q=BLOCKSPARSE_Q,
+    )
+
+
+def _create_empty_block_sparse_mask(query, key, value):
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:135]: Add a comment about how this works.

**Review by Chillee — APPROVED** [review]

File: test/inductor/test_flex_attention.py (hunk 11 above)
Comment (zou3519) [reply test/inductor/test_flex_attention.py:1168]: What exactly do you want fixed here?

Comment (Chillee) [reply test/inductor/test_flex_attention.py:1168]: Today in the IR we end up printing out all of these dead nodes that aren't used anywhere.

Comment (yanboliang) [reply test/inductor/test_flex_attention.py:1225]: Removing the unused ```new_empty``` would be addressed as one of the follow-ups.

File: torch/_dynamo/variables/higher_order_ops.py (hunk 12 above)
Comment (yanboliang) [reply torch/_dynamo/variables/higher_order_ops.py:1583]: because ```p_args``` are from ```create_wrapped_node``` which generates subgraph body node and lifted args, we should split them to follow the particular order of the HOO's arguments.

File: torch/_inductor/kernel/flex_attention.py (hunk 13 above)
Comment (yanboliang) [reply torch/_inductor/kernel/flex_attention.py]: Updated to ```SPARSE_KV_NUM_BLKS```

File: torch/_inductor/kernel/flex_attention.py (hunk 14 above)
Comment (yanboliang) [reply torch/_inductor/kernel/flex_attention.py]: It's typo, changed back.

File: test/inductor/test_flex_attention.py [hunk 16]
```diff
@@ -44,8 +46,25 @@
...
 
-def create_attention(score_mod):
-    return functools.partial(_flex_attention, score_mod=score_mod)
+def create_attention(score_mod, block_sparse_mask):
+    return functools.partial(
+        _flex_attention, score_mod=score_mod, block_sparse_mask=block_sparse_mask
+    )
+
+
+def create_block_sparse_mask_from_score_mod(score_mod, query, key, value):
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:55]: is this actually a blocksparse causal mask?

File: torch/_dynamo/variables/higher_order_ops.py (hunk 12 above)
Comment (drisspg) [reply torch/_dynamo/variables/higher_order_ops.py:1583]: can you write a comment about this, since this interleaving is pretty confusing

File: torch/nn/attention/_flex_attention.py [hunk 17]
```diff
@@ -39,11 +39,127 @@ def _identity(
     return score
 
 
+_DEFAULT_BLOCKSPARSE_SIZE = 128
+
+
+class _BlockSparseMask:
```
Comment (drisspg) [inline torch/nn/attention/_flex_attention.py:45]: were you able to make this a named tuple?

File: torch/_inductor/kernel/flex_attention.py [hunk 18]
```diff
@@ -201,6 +232,27 @@ def build_subgraph_buffer(
...
+    SPARSE_Z = {{size("SPARSE_KV_NUM_BLKS", 0)}}
+    SPARSE_H = {{size("SPARSE_KV_NUM_BLKS", 1)}}
+
+    sparse_idx_z = off_z % SPARSE_Z
+    sparse_idx_h = off_h % SPARSE_H
+
+    BLOCKSPARSE_Q_MULTIPLE: tl.constexpr = (BLOCKSPARSE_Q // BLOCK_M)
+    BLOCKSPARSE_KV_MULTIPLE: tl.constexpr = (BLOCKSPARSE_KV // BLOCK_N)
+
+    BLOCKSPARSE_Q_LEN: tl.constexpr = Q_LEN // BLOCKSPARSE_Q
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:245]: Nit: some of the naming here is a little confusing, it seems like BLOCK_SPARSE_Q is actually the Number of q Blocks?

ANd that BLOCK_SPARSE_Q is actually the sparse_Q_block_size?

File: torch/_inductor/kernel/flex_attention.py [hunk 19]
```diff
@@ -201,6 +232,27 @@ def build_subgraph_buffer(
     q_offset = off_z * stride_qz + off_h * stride_qh
     k_offset = off_z * stride_kz + off_h * stride_kh
     v_offset = off_z * stride_vz + off_h * stride_vh
+
+    SPARSE_Z = {{size("SPARSE_KV_NUM_BLKS", 0)}}
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:236]: can you update the forward and backward kernel with the args and constexprs descriptions

File: torch/nn/attention/_flex_attention.py (hunk 15 above)
Comment (drisspg) [reply torch/nn/attention/_flex_attention.py:143]: I think the comment should be as a doc string, with args, type hints, and a usage section.

Specifically if we are thinking that `_create_block_sparse_mask` is going to be the main user facing API we should explain how we expect it to use. TBH I am not even sure how it should be use and what format I would pass a mask in

Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py:143]: Oh, I think we don't plan to expose this as public. The public API should be a ```mask_func``` and we infer the block sparse mask from the mask func.

File: torch/_dynamo/variables/higher_order_ops.py (hunk 12 above)
Comment (yanboliang) [reply torch/_dynamo/variables/higher_order_ops.py:1587]: done

File: torch/_inductor/kernel/flex_attention.py (hunk 19 above)
Comment (yanboliang) [reply torch/_inductor/kernel/flex_attention.py:241]: done

File: torch/nn/attention/_flex_attention.py (hunk 17 above)
Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py:45]: I think probably yes, but I'd like to explore as a follow-up PR.

File: torch/_inductor/kernel/flex_attention.py (hunk 18 above)
Comment (yanboliang) [reply torch/_inductor/kernel/flex_attention.py]: These were updated, now all variables starting with ```SPARSE_``` means sth in the block sparse mask grid, to distinguish with the thread block grid.

File: torch/nn/attention/_flex_attention.py [hunk 20]
```diff
@@ -39,11 +39,127 @@ def _identity(
...
+        Q_BLOCK_SIZE=q_len,
+    )
+
+
 def _flex_attention(
     query: torch.Tensor,
     key: torch.Tensor,
     value: torch.Tensor,
     score_mod: _score_mod_signature = _identity,
+    block_sparse_mask: Optional[_BlockSparseMask] = None,
```
Comment (Chillee) [inline torch/nn/attention/_flex_attention.py:162]: I think we should just call this `mask` btw.

Comment (yanboliang) [reply torch/nn/attention/_flex_attention.py:162]: Will update in the follow-ups.
